File: repertoire/data-integration/providers/connector_protocol/graphql.ts
// GraphQL — connector_protocol provider
// GraphQL connector with variable binding, relay cursor pagination, query batching, and schema introspection

export interface ConnectorConfig {
  baseUrl?: string;
  connectionString?: string;
  auth?: Record<string, unknown>;
  headers?: Record<string, string>;
  options?: Record<string, unknown>;
}

export interface QuerySpec {
  path?: string;
  query?: string;
  params?: Record<string, unknown>;
  cursor?: string;
  limit?: number;
}

export interface WriteResult { created: number; updated: number; skipped: number; errors: number; }
export interface TestResult { connected: boolean; message: string; latencyMs?: number; }
export interface StreamDef { name: string; schema: Record<string, unknown>; supportedSyncModes: string[]; }
export interface DiscoveryResult { streams: StreamDef[]; }

export const PROVIDER_ID = 'graphql';
export const PLUGIN_TYPE = 'connector_protocol';

interface GraphQLResponse {
  data?: Record<string, unknown>;
  errors?: Array<{ message: string; locations?: Array<{ line: number; column: number }> }>;
}

interface RelayConnection {
  edges?: Array<{ node: Record<string, unknown>; cursor?: string }>;
  pageInfo?: { hasNextPage: boolean; endCursor?: string };
  nodes?: Record<string, unknown>[];
  totalCount?: number;
}

async function executeGraphQL(
  endpoint: string,
  query: string,
  variables: Record<string, unknown>,
  headers: Record<string, string>
): Promise<GraphQLResponse> {
  const resp = await fetch(endpoint, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json', ...headers },
    body: JSON.stringify({ query, variables }),
  });
  if (!resp.ok) throw new Error(`GraphQL request failed: ${resp.status}`);
  return resp.json();
}

async function executeBatchGraphQL(
  endpoint: string,
  queries: Array<{ query: string; variables: Record<string, unknown> }>,
  headers: Record<string, string>
): Promise<GraphQLResponse[]> {
  const resp = await fetch(endpoint, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json', ...headers },
    body: JSON.stringify(queries),
  });
  if (!resp.ok) throw new Error(`GraphQL batch request failed: ${resp.status}`);
  return resp.json();
}

const INTROSPECTION_QUERY = `
  query IntrospectionQuery {
    __schema {
      queryType { name }
      types {
        name
        kind
        fields {
          name
          type {
            name
            kind
            ofType { name kind }
          }
        }
      }
    }
  }
`;

function extractConnection(data: Record<string, unknown>): RelayConnection | null {
  for (const value of Object.values(data)) {
    if (value && typeof value === 'object' && ('edges' in (value as object) || 'nodes' in (value as object))) {
      return value as RelayConnection;
    }
    if (value && typeof value === 'object') {
      const nested = extractConnection(value as Record<string, unknown>);
      if (nested) return nested;
    }
  }
  return null;
}

function buildHeaders(config: ConnectorConfig): Record<string, string> {
  const headers: Record<string, string> = { ...config.headers };
  if (config.auth) {
    const token = config.auth.token as string | undefined;
    const style = config.auth.style as string | undefined;
    if (style === 'bearer' && token) {
      headers['Authorization'] = `Bearer ${token}`;
    } else if (style === 'api_key') {
      const headerName = (config.auth.apiKeyHeader as string) ?? 'X-API-Key';
      headers[headerName] = (config.auth.apiKey as string) ?? '';
    }
  }
  return headers;
}

export class GraphqlConnectorProvider {
  async *read(query: QuerySpec, config: ConnectorConfig): AsyncGenerator<Record<string, unknown>> {
    const endpoint = config.baseUrl ?? '';
    const gqlQuery = query.query ?? '';
    const headers = buildHeaders(config);
    const pageSize = query.limit ?? 50;
    const useRelay = (config.options?.relayPagination as boolean) ?? true;

    if (!useRelay) {
      const variables = { ...query.params, first: pageSize };
      const result = await executeGraphQL(endpoint, gqlQuery, variables, headers);
      if (result.errors?.length) throw new Error(result.errors.map(e => e.message).join('; '));
      if (!result.data) return;
      const connection = extractConnection(result.data);
      if (connection?.edges) {
        for (const edge of connection.edges) yield edge.node;
      } else if (connection?.nodes) {
        for (const node of connection.nodes) yield node as Record<string, unknown>;
      } else if (Array.isArray(Object.values(result.data)[0])) {
        for (const item of Object.values(result.data)[0] as Record<string, unknown>[]) yield item;
      }
      return;
    }

    let cursor = query.cursor ?? null;
    let hasMore = true;

    while (hasMore) {
      const variables: Record<string, unknown> = {
        ...query.params,
        first: pageSize,
        ...(cursor ? { after: cursor } : {}),
      };

      const result = await executeGraphQL(endpoint, gqlQuery, variables, headers);
      if (result.errors?.length) throw new Error(result.errors.map(e => e.message).join('; '));
      if (!result.data) break;

      const connection = extractConnection(result.data);
      if (!connection) break;

      if (connection.edges) {
        for (const edge of connection.edges) {
          yield edge.node;
        }
      } else if (connection.nodes) {
        for (const node of connection.nodes) {
          yield node as Record<string, unknown>;
        }
      }

      hasMore = connection.pageInfo?.hasNextPage ?? false;
      cursor = connection.pageInfo?.endCursor ?? null;
    }
  }

  async write(records: Record<string, unknown>[], config: ConnectorConfig): Promise<WriteResult> {
    const endpoint = config.baseUrl ?? '';
    const mutation = (config.options?.mutation as string) ?? '';
    const batchSize = (config.options?.batchSize as number) ?? 10;
    const headers = buildHeaders(config);
    const result: WriteResult = { created: 0, updated: 0, skipped: 0, errors: 0 };
    const useBatching = (config.options?.batchRequests as boolean) ?? false;

    if (useBatching) {
      for (let i = 0; i < records.length; i += batchSize) {
        const batch = records.slice(i, i + batchSize);
        const queries = batch.map(record => ({
          query: mutation,
          variables: { input: record },
        }));
        try {
          const responses = await executeBatchGraphQL(endpoint, queries, headers);
          for (const resp of responses) {
            if (resp.errors?.length) result.errors++;
            else result.created++;
          }
        } catch {
          result.errors += batch.length;
        }
      }
    } else {
      for (const record of records) {
        try {
          const resp = await executeGraphQL(endpoint, mutation, { input: record }, headers);
          if (resp.errors?.length) result.errors++;
          else result.created++;
        } catch {
          result.errors++;
        }
      }
    }
    return result;
  }

  async test(config: ConnectorConfig): Promise<TestResult> {
    const endpoint = config.baseUrl ?? '';
    const headers = buildHeaders(config);
    const start = Date.now();
    try {
      const result = await executeGraphQL(endpoint, '{ __typename }', {}, headers);
      const hasErrors = !!result.errors?.length;
      return {
        connected: !hasErrors && !!result.data,
        message: hasErrors ? result.errors!.map(e => e.message).join('; ') : 'Connected successfully',
        latencyMs: Date.now() - start,
      };
    } catch (err) {
      return { connected: false, message: String(err), latencyMs: Date.now() - start };
    }
  }

  async discover(config: ConnectorConfig): Promise<DiscoveryResult> {
    const endpoint = config.baseUrl ?? '';
    const headers = buildHeaders(config);
    try {
      const result = await executeGraphQL(endpoint, INTROSPECTION_QUERY, {}, headers);
      if (!result.data) return { streams: [] };
      const schema = result.data.__schema as {
        types: Array<{ name: string; kind: string; fields?: Array<{ name: string; type: { name: string; kind: string } }> }>;
      };
      const queryTypes = schema.types.filter(
        t => t.kind === 'OBJECT' && !t.name.startsWith('__') && t.fields && t.fields.length > 0
      );
      return {
        streams: queryTypes.map(t => ({
          name: t.name,
          schema: {
            type: 'object',
            properties: Object.fromEntries(
              (t.fields ?? []).map(f => [f.name, { type: f.type.name ?? f.type.kind }])
            ),
          },
          supportedSyncModes: ['full_refresh'],
        })),
      };
    } catch {
      return { streams: [] };
    }
  }
}

export default GraphqlConnectorProvider;

File: repertoire/data-integration/providers/connector_protocol/graphql.rs
// GraphQL — connector_protocol provider
// GraphQL connector with variable binding, relay cursor pagination, query batching, and schema introspection

use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use std::collections::HashMap;
use std::time::Instant;

pub const PROVIDER_ID: &str = "graphql";
pub const PLUGIN_TYPE: &str = "connector_protocol";

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConnectorConfig {
    pub base_url: Option<String>,
    pub connection_string: Option<String>,
    pub auth: Option<HashMap<String, String>>,
    pub headers: Option<HashMap<String, String>>,
    pub options: Option<HashMap<String, Value>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct QuerySpec {
    pub path: Option<String>,
    pub query: Option<String>,
    pub params: Option<HashMap<String, Value>>,
    pub cursor: Option<String>,
    pub limit: Option<u64>,
}

pub type Record = HashMap<String, Value>;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WriteResult { pub created: u64, pub updated: u64, pub skipped: u64, pub errors: u64 }

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TestResult { pub connected: bool, pub message: String, pub latency_ms: Option<u64> }

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StreamDef {
    pub name: String,
    pub schema: HashMap<String, Value>,
    pub supported_sync_modes: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DiscoveryResult { pub streams: Vec<StreamDef> }

#[derive(Debug)]
pub struct ConnectorError(pub String);
impl std::fmt::Display for ConnectorError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result { write!(f, "{}", self.0) }
}
impl std::error::Error for ConnectorError {}

#[derive(Debug, Deserialize)]
struct GraphQLResponse {
    data: Option<Value>,
    errors: Option<Vec<GraphQLError>>,
}

#[derive(Debug, Deserialize)]
struct GraphQLError {
    message: String,
}

#[derive(Debug, Deserialize)]
struct PageInfo {
    has_next_page: Option<bool>,
    end_cursor: Option<String>,
}

const INTROSPECTION_QUERY: &str = r#"
  query IntrospectionQuery {
    __schema {
      queryType { name }
      types {
        name kind
        fields { name type { name kind ofType { name kind } } }
      }
    }
  }
"#;

pub struct GraphqlConnectorProvider {
    config: Option<ConnectorConfig>,
    client: reqwest::Client,
}

impl GraphqlConnectorProvider {
    pub fn new() -> Self {
        Self {
            config: None,
            client: reqwest::Client::new(),
        }
    }

    fn build_headers(&self, config: &ConnectorConfig) -> HashMap<String, String> {
        let mut headers = HashMap::new();
        headers.insert("Content-Type".into(), "application/json".into());
        if let Some(cfg_h) = &config.headers {
            headers.extend(cfg_h.clone());
        }
        if let Some(auth) = &config.auth {
            match auth.get("style").map(|s| s.as_str()) {
                Some("bearer") => {
                    if let Some(token) = auth.get("token") {
                        headers.insert("Authorization".into(), format!("Bearer {}", token));
                    }
                }
                Some("api_key") => {
                    let header = auth.get("apiKeyHeader").map(|s| s.as_str()).unwrap_or("X-API-Key");
                    if let Some(key) = auth.get("apiKey") {
                        headers.insert(header.into(), key.clone());
                    }
                }
                _ => {}
            }
        }
        headers
    }

    async fn execute_graphql(
        &self,
        endpoint: &str,
        query: &str,
        variables: Value,
        headers: &HashMap<String, String>,
    ) -> Result<GraphQLResponse, ConnectorError> {
        let body = json!({ "query": query, "variables": variables });
        let mut req = self.client.post(endpoint);
        for (k, v) in headers {
            req = req.header(k.as_str(), v.as_str());
        }
        let resp = req
            .json(&body)
            .send()
            .await
            .map_err(|e| ConnectorError(e.to_string()))?;
        if !resp.status().is_success() {
            return Err(ConnectorError(format!("HTTP {}", resp.status())));
        }
        resp.json::<GraphQLResponse>()
            .await
            .map_err(|e| ConnectorError(e.to_string()))
    }

    fn extract_connection(data: &Value) -> Option<(Vec<Record>, Option<PageInfo>)> {
        if let Some(obj) = data.as_object() {
            for (_key, value) in obj {
                if let Some(inner) = value.as_object() {
                    if inner.contains_key("edges") || inner.contains_key("nodes") {
                        let mut records = Vec::new();
                        if let Some(edges) = inner.get("edges").and_then(|e| e.as_array()) {
                            for edge in edges {
                                if let Some(node) = edge.get("node") {
                                    if let Ok(rec) = serde_json::from_value(node.clone()) {
                                        records.push(rec);
                                    }
                                }
                            }
                        } else if let Some(nodes) = inner.get("nodes").and_then(|n| n.as_array()) {
                            for node in nodes {
                                if let Ok(rec) = serde_json::from_value(node.clone()) {
                                    records.push(rec);
                                }
                            }
                        }
                        let page_info = inner.get("pageInfo").and_then(|pi| {
                            Some(PageInfo {
                                has_next_page: pi.get("hasNextPage").and_then(|v| v.as_bool()),
                                end_cursor: pi.get("endCursor").and_then(|v| v.as_str()).map(String::from),
                            })
                        });
                        return Some((records, page_info));
                    }
                }
                if let Some(result) = Self::extract_connection(value) {
                    return Some(result);
                }
            }
        }
        None
    }

    pub async fn read(
        &self,
        query: &QuerySpec,
        config: &ConnectorConfig,
    ) -> Result<Vec<Record>, ConnectorError> {
        let endpoint = config.base_url.as_deref().unwrap_or("");
        let gql_query = query.query.as_deref().unwrap_or("");
        let headers = self.build_headers(config);
        let page_size = query.limit.unwrap_or(50);
        let mut all_records = Vec::new();
        let mut cursor = query.cursor.clone();
        let mut has_more = true;

        while has_more {
            let mut variables = query.params.as_ref()
                .map(|p| serde_json::to_value(p).unwrap_or(Value::Object(Default::default())))
                .unwrap_or(json!({}));
            if let Some(obj) = variables.as_object_mut() {
                obj.insert("first".into(), json!(page_size));
                if let Some(c) = &cursor {
                    obj.insert("after".into(), json!(c));
                }
            }

            let result = self.execute_graphql(endpoint, gql_query, variables, &headers).await?;
            if let Some(errors) = &result.errors {
                if !errors.is_empty() {
                    return Err(ConnectorError(errors.iter().map(|e| e.message.clone()).collect::<Vec<_>>().join("; ")));
                }
            }
            let data = match result.data {
                Some(d) => d,
                None => break,
            };

            if let Some((records, page_info)) = Self::extract_connection(&data) {
                all_records.extend(records);
                if let Some(pi) = page_info {
                    has_more = pi.has_next_page.unwrap_or(false);
                    cursor = pi.end_cursor;
                } else {
                    has_more = false;
                }
            } else {
                has_more = false;
            }
        }
        Ok(all_records)
    }

    pub async fn write(
        &self,
        records: &[Record],
        config: &ConnectorConfig,
    ) -> Result<WriteResult, ConnectorError> {
        let endpoint = config.base_url.as_deref().unwrap_or("");
        let mutation = config.options.as_ref()
            .and_then(|o| o.get("mutation"))
            .and_then(|v| v.as_str())
            .unwrap_or("");
        let headers = self.build_headers(config);
        let mut result = WriteResult { created: 0, updated: 0, skipped: 0, errors: 0 };

        for record in records {
            let variables = json!({ "input": record });
            match self.execute_graphql(endpoint, mutation, variables, &headers).await {
                Ok(resp) => {
                    if resp.errors.as_ref().map_or(false, |e| !e.is_empty()) {
                        result.errors += 1;
                    } else {
                        result.created += 1;
                    }
                }
                Err(_) => result.errors += 1,
            }
        }
        Ok(result)
    }

    pub async fn test_connection(
        &self,
        config: &ConnectorConfig,
    ) -> Result<TestResult, ConnectorError> {
        let endpoint = config.base_url.as_deref().unwrap_or("");
        let headers = self.build_headers(config);
        let start = Instant::now();
        match self.execute_graphql(endpoint, "{ __typename }", json!({}), &headers).await {
            Ok(resp) => {
                let has_err = resp.errors.as_ref().map_or(false, |e| !e.is_empty());
                Ok(TestResult {
                    connected: !has_err && resp.data.is_some(),
                    message: if has_err { "GraphQL errors returned".into() } else { "Connected successfully".into() },
                    latency_ms: Some(start.elapsed().as_millis() as u64),
                })
            }
            Err(e) => Ok(TestResult {
                connected: false,
                message: e.0,
                latency_ms: Some(start.elapsed().as_millis() as u64),
            }),
        }
    }

    pub async fn discover(
        &self,
        config: &ConnectorConfig,
    ) -> Result<DiscoveryResult, ConnectorError> {
        let endpoint = config.base_url.as_deref().unwrap_or("");
        let headers = self.build_headers(config);
        match self.execute_graphql(endpoint, INTROSPECTION_QUERY, json!({}), &headers).await {
            Ok(resp) => {
                let data = resp.data.unwrap_or(json!(null));
                let types = data.pointer("/__schema/types")
                    .and_then(|t| t.as_array())
                    .cloned()
                    .unwrap_or_default();
                let streams: Vec<StreamDef> = types.into_iter()
                    .filter(|t| {
                        let kind = t.get("kind").and_then(|k| k.as_str()).unwrap_or("");
                        let name = t.get("name").and_then(|n| n.as_str()).unwrap_or("");
                        kind == "OBJECT" && !name.starts_with("__")
                    })
                    .map(|t| {
                        let name = t.get("name").and_then(|n| n.as_str()).unwrap_or("").to_string();
                        let mut schema = HashMap::new();
                        if let Some(fields) = t.get("fields").and_then(|f| f.as_array()) {
                            for field in fields {
                                let fname = field.get("name").and_then(|n| n.as_str()).unwrap_or("");
                                let ftype = field.pointer("/type/name").and_then(|n| n.as_str()).unwrap_or("Any");
                                schema.insert(fname.to_string(), json!(ftype));
                            }
                        }
                        StreamDef {
                            name,
                            schema,
                            supported_sync_modes: vec!["full_refresh".into()],
                        }
                    })
                    .collect();
                Ok(DiscoveryResult { streams })
            }
            Err(_) => Ok(DiscoveryResult { streams: Vec::new() }),
        }
    }
}
